src/account/UpgradeableModularAccount.sol

```diff
@@ -529,10 +556,28 @@ contract UpgradeableModularAccount is
         // exists.
         for (uint256 i = 0; i < hooksLength; ++i) {
             bytes32 key = executionHooks.at(i);
-            (FunctionReference hookFunction, bool isPreHook, bool isPostHook,) = toExecutionHook(key);
+            (FunctionReference hookFunction, bool isPreHook, bool isPostHook, bool requireUOContext) =
+                toExecutionHook(key);
+
+            if (!isPackedUO && requireUOContext) {
+                revert RequireUserOperationContext();
+            }
 
             if (isPreHook) {
-                bytes memory preExecHookReturnData = _runPreExecHook(hookFunction, data);
+                bytes memory preExecHookReturnData;
+
+                if (isPackedUO) {
+                    if (requireUOContext) {
+                        preExecHookReturnData = _runPreExecHook(hookFunction, data);
+                    } else {
+                        PackedUserOperation memory uo = abi.decode(data, (PackedUserOperation));
+                        preExecHookReturnData =
+                            _runPreExecHook(hookFunction, abi.encodePacked(msg.sender, msg.value, uo.callData));
+                    }
+                } else {
+                    preExecHookReturnData =
+                        _runPreExecHook(hookFunction, abi.encodePacked(msg.sender, msg.value, data));
+                }
```

Comment:
Slight preference for not using nested if/else branches, and instead early-terminating with a `continue;` statement to reduce indent depth, but this is just stylistic. The workflow itself looks good for handling the different possible cases.

Reply:
The difference is mostly bytecode size by reducing `JUMP`s? I think this is more readable, will leave it because its RI, but production accounts should consider making this change